File: Data Structures/exercises/eulerianCycle/Haskell/DataStructures/Graph/EulerianCycle.hs
-------------------------------------------------------------------------------
-- Data Structures
-- Author: Salvi CF
-------------------------------------------------------------------------------

--module DataStructures.Graph.EulerianCycle(isEulerian, eulerianCycle) where
module DataStructures.Graph.EulerianCycle where

import DataStructures.Graph.Graph
import Data.List

--H.1)
{-
isEulerian :: Eq a => Graph a -> Bool
isEulerian g = foldr f z (vertices g)
  where
    f v acc = if (even $ degree g v) then (True && acc) else (False && acc)
    z = True
-}

-- Otra solución
isEulerian :: Eq a => Graph a -> Bool
isEulerian g = all (even) $ map (degree g) (vertices g)

-- H.2)
remove :: (Eq a) => Graph a -> (a,a) -> Graph a
remove g (v,u) = delete0 (vertices g') g'
  where
    g' = deleteEdge g (v,u)   -- grafo tras eliminar la arista
    delete0 [] graph = graph
    delete0 (v:vs) graph  | (degree graph v == 0) = delete0 vs (deleteVertex graph v)
                          | otherwise = delete0 vs graph

-- H.3)
extractCycle :: (Eq a) => Graph a -> a -> (Graph a, Path a)
extractCycle g v0 = ec g [v0]
  where
    ec graph p@(v:vs) | (s == v0) = (remove graph (v,s), s:p)
                      | otherwise = ec (remove graph (v,s)) (s : p)
      where
        (s:_) = successors graph v

-- Otra solución
{-
extractCycle :: (Eq a) => Graph a -> a -> (Graph a, Path a)
extractCycle g v0 = ec g [v0] v0
  where
    ec graph p v | (s == v0) = (remove graph (v,s), p++[s])
                 | otherwise = ec (remove graph (v,s)) (p++[s]) s
     where
       (s:_) = successors graph v
-}

-- H.4)
connectCycles :: (Eq a) => Path a -> Path a -> Path a
connectCycles [] ys = ys
connectCycles (x:xs) c@(y:ys) | (x == y) = c ++ xs
                              | otherwise = x : connectCycles xs c

-- H.5)
vertexInCommon :: Eq a => Graph a -> Path a -> a
vertexInCommon g ciclo = head [ v | v <- vertices g, c <- ciclo, (v == c) ]

-- H.6)
eulerianCycle :: Eq a => Graph a -> Path a
eulerianCycle g   | isEulerian g = ec g (head $ vertices g) []    -- grafo, vértice inicial, ciclo inicial
                  | otherwise = error "eulerianCycle: El grafo no es euleriano"
  where
    ec graph v c  | isEmpty graph = c
                  | otherwise = ec gsc vc (connectCycles c ciclo)
      where
        (gsc, ciclo) = extractCycle graph v   -- grafo sin el ciclo
        vc = vertexInCommon gsc ciclo         -- vértice común entre el grafo sin ciclo y el ciclo parcial

File: Data Structures/exercises/eulerianCycle/Java/eulerianCycle/src/dataStructures/graph/EulerianCycle.java
/**
 * Data Structures
 * @author Salvi CF
 */

package dataStructures.graph;

import dataStructures.list.*;

public class EulerianCycle<V> {
    private List<V> eCycle;

    @SuppressWarnings("unchecked")
    public EulerianCycle(Graph<V> g) {
        Graph<V> graph = (Graph<V>) g.clone();
        eCycle = eulerianCycle(graph);
    }

    public boolean isEulerian() {
        return eCycle != null;
    }

    public List<V> eulerianCycle() {
        return eCycle;
    }

    // J.1	Change "private" for "public" to test the function in main program
    private static <V> boolean isEulerian(Graph<V> g) {
        // TO DO
    	for (V v : g.vertices()){
    		if (g.degree(v) % 2 != 0){
    			return false;
    		}
    	}
        return true;
    }

    // J.2
    private static <V> void remove(Graph<V> g, V v, V u) {
        // TO DO
    	// Delete the edge
    	g.deleteEdge(v, u);	// graph without edge (u, v)
    	
    	// Delete nodes with degree 0
    	for (V vert : g.vertices()){
    		if (g.degree(vert) == 0){
    			g.deleteVertex(vert);
    		}
    	}
    }

    // J.3
    private static <V> List<V> extractCycle(Graph<V> g, V v0) {
        // TO DO
    	List<V> ciclo = new ArrayList<>();
    	ciclo.append(v0);
    	
    	V current = v0;
    	V suc = null;
    	
    	while (suc != v0){
        	for (V v : g.successors(current)){
        		suc = v;
        		break;
        	}
        	ciclo.append(suc);
        	remove(g, current, suc);
        	current = suc;
    	}
    	
        return ciclo;
    }

    // J.4
    // https://www.arquitecturajava.com/java-value-vs-reference-y-sus-curiosidades/
    private static <V> void connectCycles(List<V> xs, List<V> ys) {
    	// TO DO
    	// List<V> aux = new ArrayList<>(); xs = aux; this does not modify xs...
    	
    	if (xs.isEmpty()){
    		for (V v : ys){
    			xs.append(v);
    		}
    	} else {
    		V y = ys.get(0);
    		int idx = 0;
    		for (V v : xs){
    			if (v == y){
    				xs.remove(idx);
    				for (V u : ys){
    					xs.insert(idx, u);
    					idx++;
    				}
    				break;
    			}
    			idx++;
    		}
    	}
    }

    // J.5
    public static <V> V vertexInCommon(Graph<V> g, List<V> cycle) {
    	// TO DO
    	V res = null;
    	for (V v : g.vertices()){
    		for (V u : cycle){
    			if (v == u){
    				res = v;
    				break;
    			}
    		} 
    		if (res != null){ break; }
    	} 
    	return res;
    }

    // J.6
    private static <V> List<V> eulerianCycle(Graph<V> g) {
    	// TO DO
    	if (isEulerian(g)){
        	
        	List<V> ciclo = new ArrayList<>();
        	List<V> parcial = new ArrayList<>();
        	V vc = null;	// vertex in common
        	while (!g.isEmpty()){
            	for (V v : g.vertices()){	// extract one vertex of the graph
            		vc = v;
            		break;
            	}
            	parcial = extractCycle(g, vc);
            	connectCycles(ciclo, parcial);
            	vc = vertexInCommon(g, parcial);
        	}
        	
            return ciclo;
    	}
    	return null;
    }
}